adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala

```diff
@@ -171,99 +223,215 @@ case class GenomeBins(binSize: Long, seqLengths: Map[String, Long]) extends Seri
 
-  private def advanceCache(until: Long): Unit = {
-    while (right.hasNext && right.head._1._1.start < until) {
-      val x = right.next()
-      cache += x._1._1 -> x._2
-    }
-  }
+  // stores the current set of joined pairs
+  protected var hits: Iterator[(RT, RU)] = Iterator.empty
 
-  private def pruneCache(to: Long): Unit = {
-    cache.dropWhile(_._1.end <= to)
-  }
+  protected def advanceCache(until: Long)
+
+  protected def pruneCache(to: Long)
 
   private def getHits(): Unit = {
     // if there is nothing more in left, then I'm done
     while (left.hasNext && hits.isEmpty) {
       // there is more in left...
-      val ((nextLeftRegion, _), nextLeft) = left.next
+      val nl = left.next
+      val ((nextLeftRegion, _), nextLeft) = nl
       // ...so check whether I need to advance the cache
       // (where nextLeftRegion's end is further than prevLeftRegion's end)...
       // (the null checks are for the first iteration)
       if (prevLeftRegion == null || nextLeftRegion.end > prevLeftRegion.end) {
         advanceCache(nextLeftRegion.end)
       }
       // ...and whether I need to prune the cache
-      if (prevLeftRegion == null || nextLeftRegion.start > prevLeftRegion.start) {
+      if (prevLeftRegion == null ||
+        nextLeftRegion.start > prevLeftRegion.start) {
         pruneCache(nextLeftRegion.start)
       }
       // at this point, we effectively do a cross-product and filter; this could probably
       // be improved by making cache a fancier data structure than just a list
       // we filter for things that overlap, where at least one side of the join has a start position
       // in this partition
-      hits = cache
-        .filter(y => {
-          y._1.overlaps(nextLeftRegion) &&
-            (y._1.start >= binRegion.start || nextLeftRegion.start >= binRegion.start)
-        })
-        .map(y => (nextLeft, y._2))
-        .toList
+      hits = processHits(nextLeft, nextLeftRegion)
+
+      assert(prevLeftRegion == null ||
+        (prevLeftRegion.referenceName == nextLeftRegion.referenceName &&
+          prevLeftRegion.start < nextLeftRegion.start ||
+          (prevLeftRegion.start == nextLeftRegion.start &&
+            prevLeftRegion.end <= nextLeftRegion.end)),
+        "Left iterator in join violates sorted order invariant.")
       prevLeftRegion = nextLeftRegion
     }
   }
 
-  def hasNext: Boolean = {
+  protected def processHits(currentLeft: T,
+                            currentLeftRegion: ReferenceRegion): Iterator[(RT, RU)]
+
+  final def hasNext: Boolean = {
     // if the list of current hits is empty, try to refill it by moving forward
     if (hits.isEmpty) {
       getHits()
     }
     // if hits is still empty, I must really be at the end
-    !hits.isEmpty
+    hits.hasNext
+  }
+
+  final def next: (RT, RU) = {
+    hits.next
+  }
+}
+
+private trait VictimlessSortedIntervalPartitionJoin[T, U, RU] extends SortedIntervalPartitionJoin[T, U, T, RU] with Serializable {
```

Comment:
Some of the outer joins make use of a "victim cache" to store elements from one of the two iterators that did not match to an element in the other iterator. The phrase "victim cache" comes from a type of cache that is _occasionally_ used in computer architecture to "save" cache lines that have been evicted. I'll make a pass and add more docs.